### PR TITLE
Improve reuse of incremental parser

### PR DESCRIPTION
**Pull request by mpsijm**
Depends on metaborg/sdf#61.

TL;DR: This PR fixes a bug related to parse node reuse in the incremental parser and optimizes this fix as much as possible.

Note that this PR merges to `master`, and includes #98, which was previously merged into `develop/jsglr2`. The full Spoofax build succeeds on my machine™.

~~I'm also running a large benchmark with the merge result right now, so maybe it's best to wait with merging until that is finished, just in case. However, the code is ready for review :slightly_smiling_face:~~ Done! :smile: https://www.spoofax.dev/jsglr2evaluation-site/2021-06-01_04:18/index.html

### The long story

Commit fad7aee5d47f4e0ea28a3ae614d87f37457b7fe3 introduced a test as reproduction of some inputs that failed while benchmarking the incremental parser. In this case, one branch of the parser had already committed to reusing a parse node containing `et` (from the identifier `getMarkers`), while this parse node should actually be broken down to become part of the keyword `return`. The first fix for this was to allow the incremental parser to break down already-to-be-shifted parse nodes. However, this greatly decreased the amount of reuse, so I wanted to optimize this.

In the end, I got to the implementation as it currently stands. Most of the algorithmic decisions have been added as comments to the code of `IncrementalParser`, but if you have any more questions, feel free to ask, because then I can add more comments to make the code understandable in the future as well.

Another honorable mention goes to the incremental input stacks, which have been almost completely revamped. Before this PR, the list of `EditorUpdate`s would be applied to the parse tree of the previous version to create a tree containing temporary nodes, which would then be broken down again during the incremental parse. Now, the input stack processes the `EditorUpdate`s while parsing. Not only is this more time- and memory-efficient because we don't need temporary parse nodes anymore, but it also allowed me to implement an optimization.

The optimization starts with a test to check if "the lookahead of the lookahead is unchanged". More precisely: consider a parse node `P` that is currently at the top of the input stack (so it serves as the current "lookahead" of the parser). `P` cannot be irreusable, else the parser would break it down. Also consider the character at offset `c` that directly follows `P` (the "lookahead of the lookahead"). Now, if offset `c` is unchanged, then `P` can be reused (in other words, doesn't have to be broken down), even when there are multiple stacks "fighting" for this parse node as in the identifier-vs.-keyword scenario above. The reason why this is safe, is because `P` was created using _only_ the character at offset `c` as lookahead, since `P` was not created while multiple parse stacks were active, else it would have been irreusable.

Once the "the lookahead of the lookahead is unchanged" check fails (i.e., the character following `P` _did_ change), then this check may no longer succeed until we have passed offset `c`. This is because `P` might need to be completely restructured, which causes a different parse stack to survive by the time the parser arrives at the character following `P`.

This PR also contains miscellaneous other small fixes, but you'll find those in the commit log and in the diff.

P.S. Hurray for the 100th PR in this repository! :100: :tada:

### CODE REVIEW

**Comment by mpsijm** [comment]
The large benchmark completed with beautiful results! :smile: https://www.spoofax.dev/jsglr2evaluation-site/2021-06-01_04:18/index.html

The parse node reuse is looking perfect to me, so it seems like the parser bug has been solved without impacting reuse too much :slightly_smiling_face: 

What draws my attention is that some versions of `StringUtils.java` take quite long to parse, even though the change is small. For example, version 6 → 7 has a difference of 8 characters, but the incremental parse time is 200+ ms. I'll still investigate that further, but I have already confirmed that this slowdown isn't caused by the parser itself. Current suspects are the preceding string diff algorithm and the `Parser.complete` function that performs a full visit of the parse forest. However, that shouldn't block the merging of these PRs in my opinion.

**Comment by jasperdenkers** [comment]
Awesome 🎉
